### ISSUE TITLE
Accessing nested values is slow

### ISSUE THREAD

**Issue by Arcturuss** [dynaconf/dynaconf#844]
Accessing nested values on any level is much slower than with 'simple' keys. Minimal example below.  
Seems related to #713 but I'm not sure.

![image](https://user-images.githubusercontent.com/465493/208750669-a95f4129-939f-4f7c-9d56-c7cf81dfd2d9.png)

[Py-spy output](https://arcturuss.github.io/stuff/profile.svg)

<details>
  <summary>test.toml</summary>
    
    level0 = 123
  
    [level1]
    item = 123
    
    [level1.level2]
    item = 234
    
    [level1.level2.level3]
    item = 3452
    
    [level1.level2.level3.level4]
    item = 3452
    
    [level1.level2.level3.level4.level5]
    item = 2342
</details>

<details>
  <summary>test.py</summary>

    from dynaconf import Dynaconf

    settings = Dynaconf(settings_file="test.toml")
    
    for _ in range(10000):
        tmp = settings.level1.level2.level3.level4.level5.item
</details>

<details>
  <summary>jupyter</summary>

    from dynaconf import Dynaconf
    
    settings = Dynaconf(settings_file="test.toml")
    
    %timeit tmp = settings.level0
    %timeit tmp = settings.level1
    %timeit tmp = settings.level1.item
    %timeit tmp = settings.level1.level2
    %timeit tmp = settings.level1.level2.item
    %timeit tmp = settings.level1.level2.level3
    %timeit tmp = settings.level1.level2.level3.item
    %timeit tmp = settings.level1.level2.level3.level4
    %timeit tmp = settings.level1.level2.level3.level4.item
    %timeit tmp = settings.level1.level2.level3.level4.level5
    %timeit tmp = settings.level1.level2.level3.level4.level5.item
</details>

**Comment by q-wertz** [comment]
Having the same issue. The `boxing.py` module is the issue (having quite similar icicle plot) and takes up approximately 35% of the overall processing time of about 5 minutes...

**Comment by q-wertz** [comment]
Just as comparison

```shell
$ python3 -m timeit -n 50 "from dynaconf import Dynaconf" "settings = Dynaconf(settings_file='test.toml')" "tmp = settings.level1.level2.level3.level4.level5.item"
50 loops, best of 5: 12.2 msec per loop
$ python3 -m timeit -n 50 "from dynaconf import Dynaconf" "settings = Dynaconf(settings_file='test.toml')" "tmp = settings['level1']['level2']['level3']['level4']['level5']['item']" 
50 loops, best of 5: 11.8 msec per loop
$ python3 -m timeit -n 50 "import tomllib" "with open('test.toml', 'rb') as f:" "    settings = tomllib.load(f)" "tmp = settings['level1']['level2']['level3']['level4']['level5']['item']"
50 loops, best of 5: 106 usec per loop
```

Is it maybe possible to "fallback" on usual dict lookup when referencing using square brackets (`__getitem__`) (of course then loosing some of the fancy stuff)

**Comment by Wenzel** [comment]
Any updates on this ticket ?

I've been a happy Dynaconf user for years, but I realized today the huge performance impact that could occur when simply retrieving values from the `settings` in my application's hot paths.

This is what `py-spy top` would have to say:
![image](https://github.com/dynaconf/dynaconf/assets/964610/a1ad6a45-3030-4b98-b6d4-e300379e0d81)

Basically spending half of the app's time inside `dynaconf/base.py:__getattribute__`

To solve this we simply had to cache Dynaconf values in `__init__` as instance attribute.

I wasn't able to fully isolate the faulty behavior, but the root cause should likely be related to this ticket, although we aren't using nested attributes.

**Comment by q-wertz** [comment]
Not 100% sure but I think they are working on a rework of dynaconf (see [this](https://github.com/dynaconf/configx/tree/main) repo) which might solve this issue.
But there is not much activity in the (GitHub) project in the last few month...

**Comment by Wenzel** [comment]
Hi @rochacbruno, and thanks for your reply and your concern for this issue.
Sorry for my delayed reponse.

>can you share how exactly you did the experiment? how to run this py-spy top locally? 

I can't share code snippet as it's a private project, and I seem to fail to isolate the problem in a repo that I could share publicly.

- I'm using the [py-spy](https://github.com/benfred/py-spy) project to monitor my Python app usage
- I'm running my app in a Docker container
- py-spy only works up until `python-3.11`, so i needed to rebase my Docker image on that. (Also not compatible with `pypy`)
- I run these commands to setup py-spy
```
docker exec -ti $(docker ps -q) bash -c 'sudo -E apt-get install -y python3-dbg'
docker exec -ti $(docker ps -q) bash -c 'sudo -E pip install py-spy'
```
 And then I simply call the `top` command first, to have an overview of which function my app spends most of its time in.
```
docker exec -ti $(docker ps -q) bash -c 'sudo -E py-spy top --pid 1'
```

From there, I figured something was wrong, and I started to `dump` the stack until it landed in Dynaconf:
```
docker exec -ti $(docker ps -q) bash -c 'sudo -E py-spy dump --pid 1'
```

Which gave me the stacktrace that I pasted above.

> are your settings all top level, without any nesting?
> are you using any lazy attributes?

I'm using nesting for some of my configuration values, but the ones we are talking about here who initate these call into Dynaconf with a huge performance impact are **not nested**.

I have two `int`, and another value which is a `list`, defined with a default value given by a factory function.

Additionally, I do have a Lazy attribute declared in the Validators, however i can't find a reference where `config3_lazy` is being used.
~~~python
def factory_config2(settings, validator):
      ....
      return []

def lazy_eval_func(value, **context):
      return range(context["this"].config1_min, context["this"].config1_max+ 1)

        Validator("config1_min", cast=int, must_exist=True),
        Validator("config1_max", cast=int, must_exist=True),
        Validator("config2", default=factory_config2),
        Validator("config3_lazy", default=Lazy(empty, formatter=lazy_eval_func)),
~~~

I do see this recursive "lazy evaluation" getting called sometimes when I dump the stack:
~~~python
Thread 1 (active+gil): "MainThread"
    append (dynaconf/vendor/box/box_list.py:63)
    __init__ (dynaconf/vendor/box/box_list.py:24)
    _recursively_evaluate_lazy_format (dynaconf/utils/__init__.py:436)
    recursively_evaluate_lazy_format (dynaconf/utils/__init__.py:424)
    evaluate (dynaconf/utils/boxing.py:29)
    get (dynaconf/vendor/box/box.py:107)
    get (dynaconf/utils/boxing.py:77)
    evaluate (dynaconf/utils/boxing.py:21)
    get (dynaconf/base.py:523)
    __getattribute__ (dynaconf/base.py:324)
    __getattr__ (dynaconf/base.py:145)
~~~

Which is very suspect because the call initiating this is the one where I fetch the value of `config2` ?

I would be happy to support your investigations with more data if I can.

**Comment by rochacbruno** [comment]
The next breaking change release 4.0.0 will remove DynaBox and recursive lazy evaluations will be faster with the addition of a lazy_mapping.

This will be addressed on issue #1128 and #1129

**Comment by Wenzel** [comment]
Great, thanks for the feedback on this issue !
I will keep an eye on Dynaconf releases and see if 4.0.0 fixes this !
